Client/Frontend/Settings/SearchCustomEngineViewController.swift

```diff
@@ -0,0 +1,727 @@
+
+import Foundation
+import Static
+import Shared
+import WebKit
+import SnapKit
+import Fuzi
+import Storage
+import Data
+
+private let log = Logger.browserLogger
+
+// MARK: - SearchCustomEngineViewController
+
+class SearchCustomEngineViewController: UIViewController {
+    
+    // MARK: AddButtonType
+    
+    private enum AddButtonType {
+        case enabled
+        case disabled
+        case loading
+    }
+    
+    // MARK: Section
+    
+    private enum Section: Int, CaseIterable {
+        case url
+        case title
+    }
+    
+    // MARK: Constants
+    
+    struct Constants {
+        static let textInputRowIdentifier = "textInputRowIdentifier"
+        static let urlInputRowIdentifier = "urlInputRowIdentifier"
+        static let titleInputRowIdentifier = "titleInputRowIdentifier"
+        static let searchEngineHeaderIdentifier = "searchEngineHeaderIdentifier"
+        static let urlEntryMaxCharacterCount  = 150
+        static let titleEntryMaxCharacterCount = 50
+    }
+    
+    // MARK: Properties
+    
+    private var profile: Profile
+        
+    private var urlText: String?
+    
+    private var titleText: String?
+    
+    private var host: URL? {
+        didSet {
+            if let host = host, oldValue != host {
+                fetchSearchEngineSupportForHost(host)
+            }
+        }
+    }
+    
+    private var openSearchEngine: OpenSearchReference? {
+        didSet {
+            checkSupportAutoAddSearchEngine()
+        }
+    }
+    
+    private var isAutoAddEnabled = false
+    
+    private var dataTask: URLSessionDataTask? {
+        didSet {
+            oldValue?.cancel()
+        }
+    }
+    
+    private var fetcher: FaviconFetcher?
+    
+    fileprivate var faviconImage: UIImage?
+    
+    private lazy var spinnerView = UIActivityIndicatorView(style: .gray).then {
+        $0.hidesWhenStopped = true
+    }
+    
+    private var tableView = UITableView(frame: .zero, style: .grouped)
+    
+    // MARK: Lifecycle
+    
+    init(profile: Profile) {
+        self.profile = profile
+        
+        super.init(nibName: nil, bundle: nil)
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        title = Strings.CustomSearchEngine.customEngineNavigationTitle
+        
+        setup()
+        doLayout()
+        changeAddButton(for: .disabled)
+    }
+    
+    // MARK: Internal
+    
+    private func setup() {
+        tableView.do {
+            $0.register(URLInputTableViewCell.self, forCellReuseIdentifier: Constants.urlInputRowIdentifier)
+            $0.register(TitleInputTableViewCell.self, forCellReuseIdentifier: Constants.titleInputRowIdentifier)
+            $0.register(SearchEngineTableViewHeader.self, forHeaderFooterViewReuseIdentifier: Constants.searchEngineHeaderIdentifier)
```

Comment:
Consider using `TableViewReusable` in BraveUI

Reply:
Was thinking to add an interface for this too, smart!